subsys/bluetooth/host/gatt.c

```diff
@@ -2542,20 +2593,143 @@ int bt_gatt_notify_cb(struct bt_conn *conn,
+	gatt_notify_flush(conn);
+
+	CHECKIF(num_params == 1) {
+		/* Use the standard notification API when sending only one
+		 * notification.
+		 */
+		return -EINVAL;
+	}
+
+	CHECKIF(conn == NULL) {
+		/* Use the standard notification API to send to all connected
+		 * peers.
+		 */
+		return -EINVAL;
+	}
+
+	if (!atomic_test_bit(bt_dev.flags, BT_DEV_READY)) {
+		return -EAGAIN;
+	}
+
+	if (conn->state != BT_CONN_CONNECTED) {
+		return -ENOTCONN;
+	}
+
+#if defined(CONFIG_BT_GATT_ENFORCE_CHANGE_UNAWARE)
+	/* BLUETOOTH CORE SPECIFICATION Version 5.3
+	 * Vol 3, Part G 2.5.3 (page 1479):
+	 *
+	 * Except for a Handle Value indication for the Service Changed
+	 * characteristic, the server shall not send notifications and
+	 * indications to such a client until it becomes change-aware.
+	 */
+	if (!bt_gatt_change_aware(conn, false)) {
+		return -EAGAIN;
+	}
+#endif
+
+	/* This API guarantees an ATT_MULTIPLE_HANDLE_VALUE_NTF over the air. */
+	if (!gatt_cf_notify_multi(conn)) {
+		return -EOPNOTSUPP;
+	}
+
 	for (i = 0; i < num_params; i++) {
-		ret = bt_gatt_notify_cb(conn, &params[i]);
-		if (ret < 0) {
-			return ret;
+		/* Compute the total data length. */
+		total_len += params[i].len;
+
+		/* Confirm that the connection has the correct level of security. */
+		if (bt_gatt_check_perm(conn, params[i].attr,
+				       BT_GATT_PERM_READ_ENCRYPT |
+				       BT_GATT_PERM_READ_AUTHEN)) {
+			BT_WARN("Link is not encrypted");
+			return -EPERM;
+		}
+
+		/* The current implementation requires the same callbacks and
+		 * user_data.
+		 */
+		if ((params[0].func != params[i].func) ||
+		    (params[0].user_data != params[i].user_data)) {
+			return -EINVAL;
+		}
+
+		/* This API doesn't support passing UUIDs. */
+		if (params[i].uuid) {
+			return -EINVAL;
+		}
+
+		/* Check if the supplied handle is invalid. */
+		if (!bt_gatt_attr_get_handle(params[i].attr)) {
+			return -EINVAL;
+		}
+	}
+
+	/* PDU length is specified with a 16-bit value. */
+	if (total_len > UINT16_MAX) {
+		return -ERANGE;
+	}
+
+	/* Check there is a bearer with a high enough MTU. */
+	if (bt_att_get_mtu(conn) <
+	    (sizeof(struct bt_att_notify_mult) + total_len)) {
+		return -ERANGE;
+	}
+
+	/* Build the PDU */
+	buf = bt_att_create_pdu(conn, BT_ATT_OP_NOTIFY_MULT,
+				sizeof(struct bt_att_notify_mult) + total_len);
+	if (!buf) {
+		return -ENOMEM;
+	}
+
+	/* Register the callback. It will be called num_params times. */
+	bt_att_set_tx_meta_data(buf, params->func, params->user_data);
+	bt_att_increment_tx_meta_data_attr_count(buf, num_params-1);
+
+	for (i = 0; i < num_params; i++) {
+		data.attr = params[i].attr;
+		data.handle = bt_gatt_attr_get_handle(data.attr);
+
+		/* Check if attribute is a characteristic then adjust the
+		 * handle
+		 */
+		if (!bt_uuid_cmp(data.attr->uuid, BT_UUID_GATT_CHRC)) {
+			struct bt_gatt_chrc *chrc = data.attr->user_data;
+
+			if (!(chrc->properties & BT_GATT_CHRC_NOTIFY)) {
+				err = -EINVAL;
+				goto fail;
+			}
```

Comment:
This may be outside the scope of this PR, but isn't it kind of odd that we only do the `BT_GATT_CHRC_NOTIFY` validation if we are adjusting the handling because the attribute is a characteristic, and not when it is a characteristic value?

Reply:
fair point